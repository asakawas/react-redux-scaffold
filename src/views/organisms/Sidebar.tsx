import React, { FunctionComponent } from 'react';
import Divider from '@material-ui/core/Divider';
import FixedWidthDrawer, { FixedWidthDrawerProps } from '~/views/atoms/FixedWidthDrawer';
import DrawerCloseButton, { DrawerCloseButtonProps } from '~/views/atoms/buttons/DrawerCloseButton';

/**
 * The type of props of Sidebar.
 */
export type SidebarProps = Readonly<
  Pick<FixedWidthDrawerProps, 'drawerWidth'> &
    DrawerCloseButtonProps & {
      /** True if the sidebar is open; false otherwise. */
      open: NonNullable<FixedWidthDrawerProps['open']>;

      /** Menu items in this sidebar. */
      children: React.ReactNode;
    }
>;

const Sidebar: FunctionComponent<SidebarProps> = ({
  open,
  onDrawerCloseButtonClick,
  drawerWidth,
  children,
}) => (
  <FixedWidthDrawer variant="persistent" anchor="left" open={open} drawerWidth={drawerWidth}>
    <DrawerCloseButton onDrawerCloseButtonClick={onDrawerCloseButtonClick} />
    <Divider />
    {children}
  </FixedWidthDrawer>
);

export default React.memo(Sidebar);
