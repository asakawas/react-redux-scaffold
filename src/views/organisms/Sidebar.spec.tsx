import React from 'react';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import Sidebar from './Sidebar';

describe('Sidebar', () => {
  test('renders the menu button as expected', () => {
    const { asFragment } = render(
      <Sidebar open onDrawerCloseButtonClick={() => {}} drawerWidth={100} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows the drawer only when it's open", () => {
    {
      const renderResult = render(
        <Sidebar open onDrawerCloseButtonClick={() => {}} drawerWidth={100} />,
      );

      const drawer = renderResult.getByLabelText('drawer');
      drawer.childNodes.forEach((node) => expect(node).toBeVisible());
    }

    cleanup();

    {
      const renderResult = render(
        <Sidebar open={false} onDrawerCloseButtonClick={() => {}} drawerWidth={100} />,
      );

      const drawer = renderResult.getByLabelText('drawer');
      drawer.childNodes.forEach((node) => expect(node).not.toBeVisible());
    }
  });

  test('calls the passed handler when the drawer close button clicked', () => {
    const handler = jest.fn();
    const renderResult = render(
      <Sidebar open onDrawerCloseButtonClick={handler} drawerWidth={100} />,
    );

    const drawerCloseButton = renderResult.getByLabelText('drawer close button');
    fireEvent.click(drawerCloseButton);
    expect(handler).toHaveBeenCalledTimes(1);
    fireEvent.click(drawerCloseButton);
    expect(handler).toHaveBeenCalledTimes(2);
  });

  test("has specified drawerWidth in its drawer's style", () => {
    const handler = jest.fn();
    const renderResult = render(
      <Sidebar open onDrawerCloseButtonClick={handler} drawerWidth={123} />,
    );

    const tree = renderResult.getByLabelText('drawer');
    expect(tree).toHaveStyle('width: 123px');
  });
});
