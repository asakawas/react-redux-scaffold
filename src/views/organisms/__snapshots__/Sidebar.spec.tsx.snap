// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders the menu button as expected 1`] = `
<DocumentFragment>
  .c0 {
  width: 100px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

<div
    aria-label="drawer"
    class="MuiDrawer-root MuiDrawer-docked c0"
  >
    <div
      class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
      style="width: 100px;"
    >
      <div
        class="c1"
      >
        <button
          aria-label="drawer close button"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <hr
        class="MuiDivider-root"
      />
    </div>
  </div>
</DocumentFragment>
`;
